src/Utility.ts

```diff
@@ -87,63 +66,48 @@ export namespace Utility {
                 return resolve(port);
             });
             server.on('error', (err: Error) => {
-                return reject(new Error(err.toString()));
+                return reject(err);
             });
             server.listen(0, '127.0.0.1');
         });
     }
 
-    export async function openFileIfExists(filepath: string): Promise<boolean> {
-        const exists: boolean = await fse.pathExists(filepath);
-        if (exists) {
-            await vscode.window.showTextDocument(vscode.Uri.file(filepath), { preview: false });
-            return true;
-        } else {
-            return false;
-        }
-    }
-
     export async function getServerPort(serverXml: string): Promise<string> | undefined {
         if (!await fse.pathExists(serverXml)) {
             throw new Error(localize('tomcatExt.noserver', 'No tomcat server.'));
         }
         const xml: string = await fse.readFile(serverXml, 'utf8');
         const jsonObj: {} = await parseXml(xml);
-        return getPortFromJson(jsonObj);
+
+        if (!jsonObj || !jsonObj[Constants.SERVER]) {
+            return undefined;
+        }
+
+        let port: string | undefined;
+        const server: {} = jsonObj[Constants.SERVER];
+        const services: {}[] = server[Constants.SERVICE];
+        if (services) {
+            const service: {} = services.find((item: { $: { name: string } }) => item.$.name === Constants.CATALINA);
+            if (service && service[Constants.CONNECTOR]) {
+                const connectors: { $: {} }[] = service[Constants.CONNECTOR];
+                const connector: { $: {} } = connectors.find((item: { $: { protocol: {} } }) =>
+                    (item.$.protocol === undefined || item.$.protocol.toString().startsWith(Constants.HTTP)));
+                if (connector && connector.$) {
+                    port = connector.$[Constants.PORT];
+                }
+            }
+        }
+        return port;
```

Comment:
the getValue utility method is created for later possible usage for getting other properties from server.xml or configuration files. Currently we only care about the port. But there may be case later for other properties. So better to keep the uitility

Reply:
It is a good idea to make a utility function to get port number. The original method is especially to get Server port. That's why I move the logic back to getServerPort.